arch: x86_64: create hole for AMD HyperTransport

thomasbarrett opened the pull request:
This is the first in a small series of PRs to add support for AMD guest with >1TiB of RAM.

## Problem Overview
The AMD IOMMU [has a reserved memory region](https://github.com/torvalds/linux/blob/6c1dd1fe5d8a1d43ed96e2e0ed44a88c73c5c039/drivers/iommu/amd/iommu.c#L2517) between `0xFD_0000_0000` and `0xFF_FFFF_FFFF`. 

[As of Linux 5.4+](https://github.com/torvalds/linux/commit/9b77e5c79840fc334a5b7f770c5ab0c09dc0e028
), the VFIO framework verifies that all memory ranges provided to VFIO_IOMMU_MAP_DMA are valid.

Attempting to create a guest with >1TiB of RAM fails when VFIO devices are used during the VFIO_IOMMU_MAP_DMA step with EINVAL. 

Existing AMD guests with <1TiB of RAM and VFIO devices may experience weird behavior and / or crashes if their mmio64 BARs overlap with the HyperTransport window (although I haven't personally experienced this).

## Solution Overview
This PR adds a new hole around 1 TiB when creating x86_64 AMD hosts. There are now up to 3 non-contiguous RAM address ranges on x86_64 AMD guests instead of just 2. To support this, we replace the `(RamRange, Option<RamRange>)` tuple construct with a `Vec<RamRange>` in a few places in the codebase.

rbradford commented:
The CI failure looks relevant to this change - can you investigate?

thomasbarrett commented:
> The CI failure looks relevant to this change - can you investigate?

The failure is caused by certain OVMF behavior not playing nicely with cloud-hypervisor. OVMF seems to have its own method of calculating the expected MMIO64 location of PCI segment 0. If it doesn't match what cloud-hypervisor set, OVMF will relocate PCI mmio64 BARs. This fails under certain circumstances (specifically when using multiple PCI Segments) if the BARs are relocated outside of the _CRS window associated with segment 0.

rbradford reviewed (commented):
What about making sure `DeviceManager::start_of_device_area` is at least 1TiB? That should limit the address range for 64-bit MMIO accesses?

thomasbarrett commented:
There are problems in both cloud-hypervisor and OVMF that need to be fixed:

There are non-breaking changes required to both cloud-hypervisor and OVMF.
- For cloud-hypervisor, this PR is necessary to create the AMD HyperTransport Hole.
- For OVMF, [this](https://github.com/tianocore/edk2/pull/5037/files) PR is necessary to use the e820 table to detect RAM instead of CMOS register (which overflows at 1 TiB). 

To guarantee correct behavior at all memory sizes on AMD machines, there are also breaking changes that needs to be made in parallel to both OVMF and cloud-hypervisor to move the mmio64 window above 1 TiB on AMD as suggested by @rbradford. I would prefer to make these changes in separate (much smaller) PRs than the non-breaking changes.

I don't think that we have any high-memory CI workers, so I will post some output from the system that I am testing on.

thomasbarrett commented:
Example:

This virtual machine was created with 1408 GiB of RAM, a VFIO passthrough NVMe device, and 24 vCPUs using the OVMF branch.
```
sudo target/debug/cloud-hypervisor  \
--kernel ~/CLOUDHV.fd  \
--cpus boot=24,topology=2:12:1:1  \
--memory size=1441792M,hugepages=on,hugepage_size=2097152  \
--disk path=~/ubuntu.raw  \
--console off  \
--serial tty \
--device path=/sys/bus/pci/devices/0000:02:00.0 \ 
-v
```

Within the guest, the expected amount of memory is present (~1408 GiB). 
```
cat /proc/meminfo
MemTotal:       1452806104 kB
```

The VFIO passthrough NVMe device is working. In this case, because the amount of RAM is >1TiB, there is no overlap between the mmio64 region and the AMD HT region.
```
lsblk
NAME    MAJ:MIN RM   SIZE RO TYPE MOUNTPOINTS
loop0     7:0    0  63.9M  1 loop /snap/core20/2105
loop1     7:1    0 111.9M  1 loop /snap/lxd/24322
loop2     7:2    0  40.4M  1 loop /snap/snapd/20671
vda     252:0    0    64G  0 disk 
├─vda1  252:1    0  63.9G  0 part /
├─vda14 252:14   0     4M  0 part 
└─vda15 252:15   0   106M  0 part /boot/efi
nvme0n1 259:0    0  11.6T  0 disk 
```

jinankjain commented on arch/src/x86_64/mod.rs:
```diff
@@ -869,9 +869,58 @@ pub fn configure_vcpu(
+        // 0 GiB ~ 3GiB: memory before the gap
+        (
+            GuestAddress(0),
+            layout::MEM_32BIT_RESERVED_START.raw_value() as usize,
```
If your guest has less than 3GiB of RAM this would not work right? This would produce broken e820 map.

thomasbarrett replied:
Sorry @jinankjain, I don't think that I follow.

I fairly sure that after these regions are further processed by `MemoryManager::create_memory_regions_from_zones` and `generate_ram_ranges` function, a memory region <3 GiB region would be shrunk down to the correct size in the E820 map. 

Also, the AMD memory map in this PR is very similar to the intel memory map, especially for small guests with < 3GiB of RAM. Check out  `arch_memory_regions_default` to compare.

jinankjain replied:
Ahh right I got confused with the previous implementation of this function. It was recently changed to adapt for other scenarios.

thomasbarrett commented:
Okay, the OVMF fix for this has been merged. What is the process for updating the recommended OVMF version? Do we need to wait until the next stable release of OVMF?

rbradford commented:
> Okay, the OVMF fix for this has been merged into edk2 repo. What is the process for updating the recommended OVMF version for cloud-hypervisor @rbradford? Do we need to wait until the next stable release of OVMF?

I updated our fork and spun a new binary out of it - https://github.com/cloud-hypervisor/edk2/releases/tag/ch-ffab6553e9 the CI will pick that up on future runs.

likebreath commented:
> > Okay, the OVMF fix for this has been merged into edk2 repo. What is the process for updating the recommended OVMF version for cloud-hypervisor @rbradford? Do we need to wait until the next stable release of OVMF?
> 
> I updated our fork and spun a new binary out of it - https://github.com/cloud-hypervisor/edk2/releases/tag/ch-ffab6553e9 the CI will pick that up on future runs.

The windows guest tests are failing from most recent main build [1] (as well as my PR for v28.x backports). It seems to be related to the edk2 change.

[1] https://cloud-hypervisor-jenkins.westus.cloudapp.azure.com/blue/organizations/jenkins/cloud-hypervisor/detail/main/803/pipeline/247

thomasbarrett commented:
> > > Okay, the OVMF fix for this has been merged into edk2 repo. What is the process for updating the recommended OVMF version for cloud-hypervisor @rbradford? Do we need to wait until the next stable release of OVMF?
> > 
> > 
> > I updated our fork and spun a new binary out of it - https://github.com/cloud-hypervisor/edk2/releases/tag/ch-ffab6553e9 the CI will pick that up on future runs.
> 
> The windows guest tests are failing from most recent main build [1] (as well as my PR for v28.x backports). It seems to be related to the edk2 change.
> 
> [1] https://cloud-hypervisor-jenkins.westus.cloudapp.azure.com/blue/organizations/jenkins/cloud-hypervisor/detail/main/803/pipeline/247

Damn. 

> > > Okay, the OVMF fix for this has been merged into edk2 repo. What is the process for updating the recommended OVMF version for cloud-hypervisor @rbradford? Do we need to wait until the next stable release of OVMF?
> > 
> > 
> > I updated our fork and spun a new binary out of it - https://github.com/cloud-hypervisor/edk2/releases/tag/ch-ffab6553e9 the CI will pick that up on future runs.
> 
> The windows guest tests are failing from most recent main build [1] (as well as my PR for v28.x backports). It seems to be related to the edk2 change.
> 
> [1] https://cloud-hypervisor-jenkins.westus.cloudapp.azure.com/blue/organizations/jenkins/cloud-hypervisor/detail/main/803/pipeline/247

Urgh. I can't run the windows integration tests locally.

thomasbarrett commented:
Hmm, the code changes that I introduced to OVMF should result in identical behavior for low memory guests. The only change in behavior to OVMF was the addition of the `Use1GPageTables` flag. Perhaps this is somehow incompatible with Windows guests or kvm_hyperv.

thomasbarrett commented:
The OVMF boot-loop issue has been partially solved. The windows boot loop was caused by an commit to OVMF that occurred in mid-December rather than the 1TiB RAM changes introduced. See [this](https://github.com/cloud-hypervisor/cloud-hypervisor/issues/5942) issue for details.

thomasbarrett commented:
Rebased because I don't want to let this go stale. The main outstanding issue is deciding how to prevent assignment of MMIO regions to the HyperTransport window.

1. Move 2nd memory region from 4 GiB to 1 TiB on AMD. This seems to have been QEMU's approach.
2. Move start of MMIO region to 1TiB on AMD systems in cloud-hypervisor and OVMF. 

If possible, even though it isn't a full solution, I would prefer to merge some of the refactors that I made here to separate the Intel and AMD arch_memory_regions and generalize RAM region handling soon rather than later to make it easier to carry out-of-tree patches while we get upstream OVMF in a stable state. Thoughts @likebreath @rbradford?

thomasbarrett commented:
Closing this for now. Will re-open at a later date once upstream OVMF is stable on cloud-hypervisor
